src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Net.Mime;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public static class HttpContentJsonExtensions
+    {
+        public static Task<object?> ReadFromJsonAsync(this HttpContent content, Type type, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+            => ReadFromJsonAsyncCore(content, type, options, cancellationToken);
+
+        public static Task<T> ReadFromJsonAsync<T>(this HttpContent content, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+            => ReadFromJsonAsyncCore<T>(content, options, cancellationToken);
+
+        private static async Task<object?> ReadFromJsonAsyncCore(HttpContent content, Type type, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            using (Stream contentStream = await GetJsonStreamFromContentAsync(content).ConfigureAwait(false))
+            {
+                return await JsonSerializer.DeserializeAsync(contentStream, type, options, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private static async Task<T> ReadFromJsonAsyncCore<T>(HttpContent content, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            using (Stream contentStream = await GetJsonStreamFromContentAsync(content).ConfigureAwait(false))
+            {
+                return await JsonSerializer.DeserializeAsync<T>(contentStream, options, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private static async Task<Stream> GetJsonStreamFromContentAsync(HttpContent content)
+        {
+            ValidateMediaType(content.Headers.ContentType?.MediaType);
+            Debug.Assert(content.Headers.ContentType != null);
+
+            Encoding? sourceEncoding = GetEncoding(content.Headers.ContentType.CharSet);
```

Comment:
@rynowak should we detect BOMs here, or is going purely by explicit Content-Type charset good?

Reply:
Ideally we'd tolerate a BOM if it's present - but it's generally a mistake for a BOM to be there. I don't think CharSet will ever tell us if a BOM is present or not.

Reply:
For instance if CharSet is omitted, utf-8 can be assumed for the content types we support - so that's a case where we don't have any info.

Reply:
Our `ReadAsStringAsync()` checks CharSet first, then checks BOM if no CharSet, then assumes UTF-8 if no BOM.

I think we should use this same flow for JSON.

Reply:
@rynowak @scalablecory what is the resolution? Can we re-use get encoding logic from `ReadAsStringAsync()`?

Reply:
Wouldn't affect logic on `System.Net.Http` be too risky for release/3.1? 
This might as well only apply for master.

Reply:
As per https://github.com/dotnet/runtime/pull/33459#discussion_r396727998, we don't want to sniff a BOM on either `application/json` or `application/<subtype>+json` content since that's against the spec.

Keeping this logic separate form `ReadAsStringAsync`.

In `ReadAsStringAsync`, we could separate the code that determines the encoding from the `charset`and the code that gets the encoding form the BOM and then re-use the first in here but as pointed on my previous comment that could be addressed on a separate PR since it applies to 5.0 only.

Reply:
That the spec forbids writing a BOM but does not forbid reading one:

> Implementations MUST NOT add a byte order mark (U+FEFF) to the beginning of a networked-transmitted JSON text.  In the interests of interoperability, implementations that parse JSON texts MAY ignore the presence of a byte order mark rather than treating it as an error.

I am fine leaving this off for now if we're willing to assume BOM usage is not a common scenario we want compat with.